File: server/pipeline/pattern_matcher.py
from __future__ import annotations

import re

from models.profile import CandidateProfile


def _parse_age(age_str: str | None) -> int | None:
    """Parse age from string like '44歳' → 44."""
    if not age_str:
        return None
    match = re.search(r"(\d+)", age_str)
    return int(match.group(1)) if match else None


def _parse_experience_years(exp_str: str | None) -> int | None:
    """Parse experience years from string.

    '10年以上' → 10, '3年' → 3, '6〜9年' → 6, '1年未満' → 0, empty → None
    """
    if not exp_str or exp_str.strip() in ("", "未入力", "なし"):
        return None
    # Match first number in the string
    match = re.search(r"(\d+)", exp_str)
    if not match:
        return None
    years = int(match.group(1))
    # "1年未満" → treat as 0
    if "未満" in exp_str and years <= 1:
        return 0
    return years


def _determine_employment_state(profile: CandidateProfile) -> str:
    """Determine employment state: 就業中, 離職中, or 在学中."""
    status = profile.employment_status or ""
    if "在学中" in status:
        return "在学中"
    if "就業中" in status:
        return "就業中"
    return "離職中"


def _determine_age_bracket(age: int | None) -> str:
    """Determine age bracket for pattern matrix.

    Returns: '40s+', 'late_30s', 'young', or 'student' (handled separately).
    """
    if age is None:
        # Default to young if unknown
        return "young"
    if age >= 40:
        return "40s+"
    if age >= 35:
        return "late_30s"
    return "young"


def _select_pattern_type(
    age_bracket: str,
    experience_years: int | None,
    employment_state: str,
) -> str:
    """Select pattern type from the age x experience matrix."""
    if employment_state == "在学中":
        return "G"

    has_exp = experience_years is not None and experience_years > 0

    if has_exp:
        if experience_years >= 10:
            if age_bracket in ("40s+", "late_30s"):
                return "A"
            else:
                return "B1"
        elif experience_years >= 6:
            return "B1"
        elif experience_years >= 3:
            return "B2"
        elif experience_years >= 1:
            if age_bracket in ("40s+", "late_30s"):
                return "C"
            else:
                return "E"
        else:
            # experience_years == 0 (未満)
            if employment_state == "就業中":
                return "F_就業中" if age_bracket == "young" else "D_就業中"
            else:
                return "F_離職中" if age_bracket == "young" else "D_離職中"
    else:
        # No experience data
        if age_bracket in ("40s+", "late_30s"):
            if employment_state == "就業中":
                return "D_就業中"
            else:
                return "D_離職中"
        else:
            if employment_state == "就業中":
                return "F_就業中"
            else:
                return "F_離職中"


def should_use_pattern(profile: CandidateProfile) -> bool:
    """Returns True if candidate should use pattern matching (no substantial work history/self_pr).

    Pattern matching is used when the candidate has no meaningful work history
    or self-PR to personalize from.
    """
    has_work_history = bool(
        profile.work_history_summary
        and profile.work_history_summary.strip()
        and profile.work_history_summary.strip() not in ("未入力", "なし", "-", "ー")
    )
    has_self_pr = bool(
        profile.self_pr
        and profile.self_pr.strip()
        and profile.self_pr.strip() not in ("未入力", "なし", "-", "ー")
    )

    return not has_work_history and not has_self_pr


def _find_pattern(
    pattern_type: str,
    patterns: list[dict],
    employment_state: str,
) -> dict | None:
    """Find the best matching pattern from the patterns list.

    Patterns may have employment_variant to distinguish 就業中/離職中 variants.
    """
    # First try exact match including employment variant suffix
    for p in patterns:
        if p["pattern_type"] == pattern_type:
            variant = p.get("employment_variant")
            if variant is None:
                return p
            if variant == employment_state:
                return p

    # Try base pattern type (strip employment suffix)
    base_type = pattern_type.split("_")[0]
    for p in patterns:
        if p["pattern_type"] == base_type:
            variant = p.get("employment_variant")
            if variant is None:
                return p
            if variant == employment_state:
                return p

    # Fallback: any pattern with matching base
    for p in patterns:
        if p["pattern_type"].startswith(base_type):
            return p

    return None


def _apply_qualification_modifier(
    text: str,
    qualifications: str,
    qualification_modifiers: list[dict],
) -> str:
    """Apply qualification-based text modifier if applicable.

    qualification_modifiers items have:
        qualification_combo: list of qualification keywords that must ALL be present
        replacement_text: text to replace a placeholder or append
    """
    if not qualification_modifiers or not qualifications:
        return text

    for modifier in qualification_modifiers:
        combo = modifier.get("qualification_combo", [])
        if all(q in qualifications for q in combo):
            replacement = modifier.get("replacement_text", "")
            if replacement:
                # Replace {資格修飾} placeholder if present, otherwise prepend
                if "{資格修飾}" in text:
                    text = text.replace("{資格修飾}", replacement)
                else:
                    text = replacement + text
            break

    # Clean up unused placeholder
    text = text.replace("{資格修飾}", "")
    return text


def match_pattern(
    profile: CandidateProfile,
    patterns: list[dict],
    qualification_modifiers: list[dict],
    feature_rotation_index: int = 0,
) -> tuple[str, str, str]:
    """Match a candidate to a pattern and generate personalized text.

    Args:
        profile: Candidate profile.
        patterns: List of pattern definitions with pattern_type, template_text,
            feature_variations, and optional employment_variant.
        qualification_modifiers: List of qualification-based text modifiers.
        feature_rotation_index: Index for rotating through feature variations.

    Returns:
        Tuple of (pattern_type, personalized_text, debug_info).

    Raises:
        ValueError: If no matching pattern is found.
    """
    age = _parse_age(profile.age)
    experience_years = _parse_experience_years(profile.experience_years)
    employment_state = _determine_employment_state(profile)
    age_bracket = _determine_age_bracket(age)

    pattern_type = _select_pattern_type(age_bracket, experience_years, employment_state)

    debug_parts = [
        f"age={age}({age_bracket})",
        f"exp={experience_years}",
        f"status={employment_state}",
        f"pattern={pattern_type}",
    ]

    pattern = _find_pattern(pattern_type, patterns, employment_state)
    if pattern is None:
        raise ValueError(
            f"パターン '{pattern_type}' が見つかりません "
            f"(age={age}, exp={experience_years}, status={employment_state})"
        )

    template_text = pattern["template_text"]

    # Select feature variation by rotation
    feature_variations = pattern.get("feature_variations", [])
    if feature_variations:
        idx = feature_rotation_index % len(feature_variations)
        feature = feature_variations[idx]
        debug_parts.append(f"feature_idx={idx}/{len(feature_variations)}")
    else:
        feature = ""

    # Replace placeholders
    personalized = template_text.replace("{特色}", feature)

    # Replace {N} with experience years if available
    if experience_years is not None and experience_years > 0:
        personalized = personalized.replace("{N}", str(experience_years))
    else:
        personalized = personalized.replace("{N}", "")

    # Replace {職種名} with qualification-derived job name
    job_name = _resolve_job_name(profile.qualifications or "")
    personalized = personalized.replace("{職種名}", job_name)

    # Apply qualification modifiers
    personalized = _apply_qualification_modifier(
        personalized,
        profile.qualifications or "",
        qualification_modifiers,
    )

    debug_info = ", ".join(debug_parts)
    return pattern_type, personalized.strip(), debug_info


def _resolve_job_name(qualifications: str) -> str:
    """Resolve primary job name from qualifications string."""
    if "看護師" in qualifications:
        return "看護師"
    if "理学療法士" in qualifications:
        return "理学療法士"
    if "言語聴覚士" in qualifications:
        return "言語聴覚士"
    if "作業療法士" in qualifications:
        return "作業療法士"
    if "医療事務" in qualifications:
        return "医療事務"
    return ""

File: extension-scout/src/shared/pattern-matcher.ts
/**
 * 型はめ（pattern matching）— サーバなしで候補者プロフィールから
 * パーソナライズ文を組み立てる。
 *
 * 等価性の拘束:
 *   Python 版 (server/pipeline/pattern_matcher.py) と挙動を一致させる。
 *   既存の Python テスト (server/tests/test_pattern_matcher.py) が
 *   基準仕様。test/pattern-matcher.test.ts が TS 側の再現性を保証する。
 *
 * この拡張（ウェルミー等）は全部型はめでサーバを呼ばないため、
 * ここが生成の中核になる。
 */

import type { CandidateProfile } from './types';

export type AgeBracket = '40s+' | 'late_30s' | 'young';
export type EmploymentState = '就業中' | '離職中' | '在学中';

export interface MatchRule {
  employment?: EmploymentState;
  age_group?: AgeBracket | null;
  exp_min?: number | null;
  exp_max?: number | null;
}

export interface Pattern {
  pattern_type: string;
  template_text: string;
  feature_variations?: string[];
  employment_variant?: EmploymentState;
  match_rules?: MatchRule[];
  job_category?: string;
}

export interface QualificationModifier {
  qualification_combo: string[];
  replacement_text: string;
}

/** "44歳" → 44, 空 → null */
export function parseAge(ageStr: string | undefined | null): number | null {
  if (!ageStr) return null;
  const m = ageStr.match(/(\d+)/);
  return m ? parseInt(m[1], 10) : null;
}

/**
 * "10年以上" → 10, "3年" → 3, "6〜9年" → 6, "1年未満" → 0,
 * 空・"未入力"・"なし" → null
 */
export function parseExperienceYears(expStr: string | undefined | null): number | null {
  if (!expStr) return null;
  const trimmed = expStr.trim();
  if (trimmed === '' || trimmed === '未入力' || trimmed === 'なし') return null;
  const m = trimmed.match(/(\d+)/);
  if (!m) return null;
  const years = parseInt(m[1], 10);
  if (trimmed.includes('未満') && years <= 1) return 0;
  return years;
}

export function determineEmploymentState(profile: CandidateProfile): EmploymentState {
  const status = profile.employment_status || '';
  if (status.includes('在学中')) return '在学中';
  if (status.includes('就業中')) return '就業中';
  return '離職中';
}

export function determineAgeBracket(age: number | null): AgeBracket {
  if (age === null) return 'young';
  if (age >= 40) return '40s+';
  if (age >= 35) return 'late_30s';
  return 'young';
}

/**
 * ハードコードされた年齢×経験年数マトリクスでパターン型を決定する。
 * match_rules がパターン定義にあればそちらが優先（_selectPatternTypeFromRules）。
 */
export function selectPatternType(
  ageBracket: AgeBracket,
  experienceYears: number | null,
  employmentState: EmploymentState
): string {
  if (employmentState === '在学中') return 'G';

  const hasExp = experienceYears !== null && experienceYears > 0;

  if (hasExp) {
    const exp = experienceYears as number;
    if (exp >= 10) {
      return ageBracket === '40s+' || ageBracket === 'late_30s' ? 'A' : 'B1';
    }
    if (exp >= 6) return 'B1';
    if (exp >= 3) return 'B2';
    if (exp >= 1) {
      return ageBracket === '40s+' || ageBracket === 'late_30s' ? 'C' : 'E';
    }
    // experience_years == 0 (未満)
    if (employmentState === '就業中') {
      return ageBracket === 'young' ? 'F_就業中' : 'D_就業中';
    }
    return ageBracket === 'young' ? 'F_離職中' : 'D_離職中';
  }

  // No experience data
  if (ageBracket === '40s+' || ageBracket === 'late_30s') {
    return employmentState === '就業中' ? 'D_就業中' : 'D_離職中';
  }
  return employmentState === '就業中' ? 'F_就業中' : 'F_離職中';
}

const INVALID_TEXT_VALUES = new Set(['未入力', 'なし', '-', 'ー', '']);

function isMeaningful(value: string | undefined | null): boolean {
  if (!value) return false;
  const t = value.trim();
  if (!t) return false;
  return !INVALID_TEXT_VALUES.has(t);
}

/**
 * パターン経路を使うべきか。work_history/self_pr が両方空なら true。
 * WelMe 等の他媒体では常に true になる前提。
 */
export function shouldUsePattern(profile: CandidateProfile): boolean {
  return !isMeaningful(profile.work_history_summary) && !isMeaningful(profile.self_pr);
}

function matchRule(
  rule: MatchRule,
  ageBracket: AgeBracket,
  experienceYears: number | null,
  employmentState: EmploymentState
): boolean {
  if (rule.employment && employmentState !== rule.employment) return false;
  if (rule.age_group && ageBracket !== rule.age_group) return false;

  const hasExpMin = 'exp_min' in rule;
  const hasExpMax = 'exp_max' in rule;

  // "exp_min": null → matches when experience_years is None (no data)
  if (hasExpMin && rule.exp_min === null && !hasExpMax) {
    return experienceYears === null;
  }

  if (rule.exp_min !== undefined && rule.exp_min !== null) {
    if (experienceYears === null || experienceYears < rule.exp_min) return false;
  }
  if (rule.exp_max !== undefined && rule.exp_max !== null) {
    if (experienceYears === null) return false;
    if (experienceYears > rule.exp_max) return false;
  }
  return true;
}

function selectPatternTypeFromRules(
  patterns: Pattern[],
  ageBracket: AgeBracket,
  experienceYears: number | null,
  employmentState: EmploymentState
): string | null {
  const hasAnyRules = patterns.some((p) => p.match_rules && p.match_rules.length > 0);
  if (!hasAnyRules) return null;

  for (const p of patterns) {
    if (!p.match_rules || p.match_rules.length === 0) continue;
    for (const rule of p.match_rules) {
      if (matchRule(rule, ageBracket, experienceYears, employmentState)) {
        if (p.employment_variant) return `${p.pattern_type}_${p.employment_variant}`;
        if (
          (employmentState === '就業中' || employmentState === '離職中') &&
          (p.pattern_type === 'D' || p.pattern_type === 'F')
        ) {
          return `${p.pattern_type}_${employmentState}`;
        }
        return p.pattern_type;
      }
    }
  }
  return null;
}

function findPattern(
  patternType: string,
  patterns: Pattern[],
  employmentState: EmploymentState
): Pattern | null {
  // 1. Exact match including employment variant suffix
  for (const p of patterns) {
    if (p.pattern_type === patternType) {
      if (!p.employment_variant) return p;
      if (p.employment_variant === employmentState) return p;
    }
  }

  // 2. Base pattern type (strip employment suffix)
  const baseType = patternType.split('_')[0];
  for (const p of patterns) {
    if (p.pattern_type === baseType) {
      if (!p.employment_variant) return p;
      if (p.employment_variant === employmentState) return p;
    }
  }

  // 3. Fallback: any pattern with matching base
  for (const p of patterns) {
    if (p.pattern_type.startsWith(baseType)) return p;
  }

  return null;
}

function applyQualificationModifier(
  text: string,
  qualifications: string,
  modifiers: QualificationModifier[]
): string {
  if (!modifiers || modifiers.length === 0 || !qualifications) {
    return text.replace('{資格修飾}', '');
  }
  let modified = text;
  for (const m of modifiers) {
    if (m.qualification_combo.every((q) => qualifications.includes(q))) {
      const replacement = m.replacement_text || '';
      if (replacement) {
        if (modified.includes('{資格修飾}')) {
          modified = modified.replace('{資格修飾}', replacement);
        } else {
          modified = replacement + modified;
        }
      }
      break;
    }
  }
  return modified.replace('{資格修飾}', '');
}

/** 資格文字列から主職種名を解決する。 */
export function resolveJobName(qualifications: string): string {
  if (qualifications.includes('看護師')) return '看護師';
  if (qualifications.includes('理学療法士')) return '理学療法士';
  if (qualifications.includes('言語聴覚士')) return '言語聴覚士';
  if (qualifications.includes('作業療法士')) return '作業療法士';
  if (qualifications.includes('医療事務')) return '医療事務';
  return '';
}

/**
 * 候補者にパターンを当てはめてパーソナライズ文を生成する。
 * @returns [pattern_type, personalized_text, debug_info]
 * @throws 該当パターンが存在しないとき
 */
export function matchPattern(
  profile: CandidateProfile,
  patterns: Pattern[],
  qualificationModifiers: QualificationModifier[] = [],
  featureRotationIndex = 0
): [string, string, string] {
  const age = parseAge(profile.age);
  const experienceYears = parseExperienceYears(profile.experience_years);
  const employmentState = determineEmploymentState(profile);
  const ageBracket = determineAgeBracket(age);

  let patternType = selectPatternTypeFromRules(
    patterns,
    ageBracket,
    experienceYears,
    employmentState
  );
  if (patternType === null) {
    patternType = selectPatternType(ageBracket, experienceYears, employmentState);
  }

  const debugParts = [
    `age=${age}(${ageBracket})`,
    `exp=${experienceYears}`,
    `status=${employmentState}`,
    `pattern=${patternType}`,
  ];

  const pattern = findPattern(patternType, patterns, employmentState);
  if (pattern === null) {
    throw new Error(
      `パターン '${patternType}' が見つかりません (age=${age}, exp=${experienceYears}, status=${employmentState})`
    );
  }

  let personalized = pattern.template_text;
  const variations = pattern.feature_variations || [];
  let feature = '';
  if (variations.length > 0) {
    const idx = featureRotationIndex % variations.length;
    feature = variations[idx];
    debugParts.push(`feature_idx=${idx}/${variations.length}`);
  }
  personalized = personalized.replace('{特色}', feature);

  if (experienceYears !== null && experienceYears > 0) {
    personalized = personalized.replace('{N}', String(experienceYears));
  } else {
    personalized = personalized.replace('{N}', '');
  }

  const jobName = resolveJobName(profile.qualifications || '');
  personalized = personalized.replace('{職種名}', jobName);

  personalized = applyQualificationModifier(
    personalized,
    profile.qualifications || '',
    qualificationModifiers
  );

  return [patternType, personalized.trim(), debugParts.join(', ')];
}
